### PR TITLE
Implement invocation strategy config

### PR DESCRIPTION
**Pull request by Veykril**
Fixes https://github.com/rust-lang/rust-analyzer/issues/10793

This allows to change how we run build scripts (and `checkOnSave`), exposing two configs:
- `once`: run the specified command once in the project root (the working dir of the server)
- `per_workspace`: run the specified command per workspace in the corresponding workspace

This also applies to `checkOnSave` likewise, though `once_in_root` is useless there currently, due to https://github.com/rust-lang/cargo/issues/11007

### CODE REVIEW

**Comment by jyn514** [comment]
@Veykril the "OnceInRoot" strategy looks like exactly what we want :) I'd love to see this merged, is there any way I can help out? Maybe test this locally on rust-lang/rust for you? Not sure how to do that but I'm happy to try.

**Comment by Veykril** [comment]
There is nothing specific blocking this PR aside from me finishing iirc. The problem with `OnceInRoot` is that checkOnSave can't really support it currently due to how diagnostics work, I'll have to look into that a bit, but for a tool to work properly for checkOnSave with `OnceInRoot` it has to prefix all the file paths in diagnostics such that they are relative to the project root (cargo currently always just emits `src/file/path/...` which r-a has to manually prefix the workspace to, but that obviously doesn't work if we don't run this per workspace)

**Comment by Veykril** [comment]
Testing it would be great as well though (I don't really know the workspace setup in rustc to reliable know how to check this right now). In theory to test that, all you need to do is set the `cargo.buildScripts.invocationStrategy` to `"once_in_root"` and set the other buildScript stuff as usual for the rustc repo. I assume the x.py script should emit the json metadata for all workspaces already in which case it should just work.

**Comment by Veykril** [comment]
Okay, the PR is basically done now, *but* checkOnSave while implemented, won't work without changes to x.py or cargo. The build script stuff should just work assuming x.py emits all the metadata for the workspaces when building with proper paths (this is an assumption, we'd need to test that). checkOnSave will not work as is though, cargo emits paths in diagnostics relative to the workspace root, that means if x.py forwards the diagnostics as is to r-a, r-a won't know what files to associate these with, because it now gets diagnostics with workspace relative paths for all workspaces without the corresponding workspaces, so either x.py has to fix these paths up (as it should know from what workspaces the diagnostics come) or we need to fix https://github.com/rust-lang/cargo/issues/11007 assuming x.py uses that under the hood. https://github.com/rust-lang/cargo/issues/11007 needs to be fixed nevertheless for `per_workspace_with_manifest_path` to work  as well.

**Comment by RalfJung** [comment]
> run the specified command per workspace in the project root with --manifest-path path/to/Cargo.toml passed to the arguments

So you decided against adding interpolation for this? That seemed like the more general thing.

> checkOnSave will not work as is though, cargo emits paths in diagnostics relative to the workspace root

In "per workspace in the project root" mode, we could declare that RA interprets diagnostics paths relative to the folder containing the workspace tom file -- that should work, right?

"once in the project root" indeed cannot currently work due to https://github.com/rust-lang/cargo/issues/11007 (unless we teach bootstrap to translate the paths it receives from cargo).

**Comment by Veykril** [comment]
> So you decided against adding interpolation for this? That seemed like the more general thing.

Not strictly, this was just simpler to implement for now. I wanted to get the general thing working first, it probably makes sense to switch this to intepolation before finalizing the PR.

> In "per workspace in the project root" mode, we could declare that RA interprets diagnostics paths relative to the folder containing the workspace toml file -- that should work, right?

Right, that mode actually works fine as is, not sure why I thought differently, there is no problem here since we still know the workspaces.

**Comment by jyn514** [comment]
> Testing it would be great as well though (I don't really know the workspace setup in rustc to reliable know how to check this right now). In theory to test that, all you need to do is set the `cargo.buildScripts.invocationStrategy` to `"once_in_root"` and set the other buildScript stuff as usual for the rustc repo. I assume the x.py script should emit the json metadata for all workspaces already in which case it should just work.

Great :) how can I use this build of rust-analyzer with rust-lang/rust? I see https://github.com/rust-lang/rust-analyzer/tree/master/docs/dev#launching-rust-analyzer, but I tried `cargo xtask install --server` and updating `rust-analyzer.server.path` but it didn't seem to do anything.

**Comment by Veykril** [comment]
Hmm, odd, that should be enough. You could also try building via `cargo build -p rust-analyzer --release` and point `rust-analyzer.server.path` to the build in your target then. If you are using VSCode you can check if you are using the self-build server by running the `show RA version` command (accesible by default via `ctrl + shift + P`), if that says version `0.0.0` it is using your custom build.

**Comment by Veykril** [comment]
Alright this is now ready (the interpolation implemented here is just a replacement of `$manifest_path`, doing a proper system can follow as that's all that's needed for now).

There is basically only the question of naming left, what should the configs be called, current names are `once_in_root` and `per_workspace`, I think these are fine but I am open to more ideas.

More important is the name for the interpolation var `$manifest_path`, should we go with snake case? SCREAMING_SNAKE_CASE? are bare name fines, or should we have some kind of namespaced system in place. Getting this right from the start would be nice, as we don't really want to have to support this special case otherwise. Alternatively, the interpolation thing can be left out of the PR, but iirc the interpolation stuff is already enough for miri to work properly right @RalfJung?


I also tested that `once_in_root` works with `linked_projects` in the rust repo. `checkOnSave` obviously doesn't as `once_in_root` is "broken" for cargo diagnostics due to https://github.com/rust-lang/cargo/issues/11007

**Comment by RalfJung** [comment]
> Alternatively, the interpolation thing can be left out of the PR, but iirc the interpolation stuff is already enough for miri to work properly right @RalfJung?

Miri is fine even without this PR, since we introduced a `./miri clippy` wrapper.

But rustc needs something like this PR -- specifically `per_workspace` with the working dir being the project root. It doesn't actually need the `manifest_path` interpolation variable though, so tying that to the working dir is a bit unfortunate for rustc. The approach I suggested for rustc would involve doing `./x.py check compiler/rustc` for *each* project, each time in the project root. Since the build is cached, re-running the command for each project is not a big deal. Basically what will happen is that when invoking this for the bootstrap project, RA will see the diagnostics for all projects but will discard most of them since they are for the wrong files. Then RA calls this again for the main compiler, gets the same diagnostics (hopefully, we need to ensure `x.py` re-emits diagnostics) and finds the diagnostics for the main workspace. (And then maybe another round for the cranelift workspace.)

Not sure how much sense that makes but it could work. I am just not sure if `x.py` really re-emits diagnostics properly.

**Comment by Veykril** [comment]
Ye, tying the working directory to the interpolation var seems rather odd, I realized this as well now. Will probably split the `per_workspace` one into two with just the working dir differing between them. And then remove the interpolation stuff for now

> But rustc needs something like this PR -- specifically per_workspace with the working dir being the project root.

Ye, the diagnostics/checkOnSave stuff the bootstrap people will probably have to figure out, this PR should give them all the tools needed for that I believe. `once_in_root` works properly for the build scripts at least so there is nothing that needs fixing there.

**Comment by RalfJung** [comment]
per_workspace_from_root without interpolation is a bit odd though since it means the script has no way to know which project even needs to be built. Granted, x.py can't (currently) make use of that information, but in general it seems like that would be good to have...

**Comment by Veykril** [comment]
Well, it would be certainly great if x.py could make use of the manifest-path arg, because a check right now is more expensive than it needs to be I think. At least from a quick test running x.py check takes a long time even if I'm only interested in one workspace (though I have almost no experience with the rustc codebase).

I'm a bit split now though, you are right that `per_workspace_in_root` is rather pointless, so the tying of the interpolation to the working dir does make somewhat sense :/ Guess I'll think some more about the general interpolation syntax then

File: crates/flycheck/src/lib.rs [hunk 1]
```diff
@@ -284,18 +299,35 @@ impl FlycheckActor {
...
         };
-        cmd.current_dir(&self.workspace_root);
+        if let InvocationStrategy::PerWorkspace = invocation_strategy {
+            let mut with_manifest_path = false;
+            for arg in args {
+                if let Some(_) = arg.find("$manifest_path") {
+                    with_manifest_path = true;
+                    cmd.arg(arg.replace(
+                        "$manifest_path",
+                        &self.root.join("Cargo.toml").display().to_string(),
```
Comment (googleson78) [inline crates/flycheck/src/lib.rs:318]: There should be a check for `rust-project.json` here, I think (and of course placing `rust-project.json` when it's being used instead)

**Comment by Veykril** [comment]
I removed the simplistic interpolation from this, as I'll rather do that properly with a bit more thought behind. The PR itself is good to go though and the rustc repo at least benefits from the `once` setting for build scripts (and technically `checkOnSave` if the x.py script fixes up the paths)
@bors merge

**Comment by ShayBox** [comment]
Could someone clarify what these settings do exactly? the descriptions are pretty vague imo and I don't understand them.

**Pull request merged**